contracts/MarginCalculator.sol

```diff
@@ -528,10 +1067,163 @@ contract MarginCalculator {
+        );
+    }
+
+    /**
+     * @dev if there is a short option and a long option in the vault, ensure that the long option is able to be used as collateral for the short option
+     * @param _vault the vault to check
+     * @param _vaultDetails vault details struct
+     * @return true if long is marginable or false if not
+     */
+    function _isMarginableLong(MarginVault.Vault memory _vault, VaultDetails memory _vaultDetails)
+        internal
+        pure
+        returns (bool)
+    {
+        // if vault is missing a long or a short, return True
+        if (!_vaultDetails.hasLong || !_vaultDetails.hasShort) return true;
+
+        if (_vaultDetails.vaultType == 1)
+            require(!_vaultDetails.hasLong, "MarginCalculator: naked margin vault cannot have long otoken");
+
+        return
+            _vault.longOtokens[0] != _vault.shortOtokens[0] &&
+            _vaultDetails.longUnderlyingAsset == _vaultDetails.shortUnderlyingAsset &&
+            _vaultDetails.longStrikeAsset == _vaultDetails.shortStrikeAsset &&
+            _vaultDetails.longCollateralAsset == _vaultDetails.shortCollateralAsset &&
+            _vaultDetails.longExpiryTimestamp == _vaultDetails.shortExpiryTimestamp &&
+            _vaultDetails.isLongPut == _vaultDetails.isShortPut;
+    }
+
+    /**
+     * @dev if there is short option and collateral asset in the vault, ensure that the collateral asset is valid for the short option
+     * @param _vault the vault to check
+     * @param _vaultDetails vault details struct
+     * @return true if marginable or false
+     */
+    function _isMarginableCollateral(MarginVault.Vault memory _vault, VaultDetails memory _vaultDetails)
+        internal
+        pure
+        returns (bool)
+    {
+        bool isMarginable = true;
+
+        if (!_vaultDetails.hasCollateral) return isMarginable;
+
+        if (_vaultDetails.hasShort) {
+            isMarginable = _vaultDetails.shortCollateralAsset == _vault.collateralAssets[0];
+        } else if (_vaultDetails.hasLong) {
+            isMarginable = _vaultDetails.longCollateralAsset == _vault.collateralAssets[0];
+        }
+
+        return isMarginable;
+    }
+
+    /**
+     * @notice get a product hash
+     * @param _underlying option underlying asset
+     * @param _strike option strike asset
+     * @param _collateral option collateral asset
+     * @param _isPut option type
+     * @return product hash
+     */
+    function _getProductHash(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut
+    ) internal pure returns (bytes32) {
+        return keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+    }
+
+    /**
+     * @notice get option cash value
+     * @dev this assume that the underlying price is denominated in strike asset
+     * for put, cv = strike price - underlying price
+     * for call, cv = underlting price - strike price
```

Comment:
underlying
prb should use below for clarity:
max(underlying price - strike price, 0)

same for puts